DigitalLearningSolutions.Data/Services/RecommendedLearningService.cs

```diff
@@ -67,40 +67,119 @@ int delegateId
             var delegateLearningLogItems = learningLogItemsDataService.GetLearningLogItems(delegateId);
 
             var recommendedResources = resources.ResourceReferences.Select(
-                rr =>
-                {
-                    var learningLogItemsForResource = delegateLearningLogItems.Where(
-                        ll => ll.ArchivedDate == null && ll.LearningHubResourceReferenceId == rr.RefId
-                    ).ToList();
-                    var incompleteLearningLogItem =
-                        learningLogItemsForResource.SingleOrDefault(ll => ll.CompletedDate == null);
-                    return new RecommendedResource(
-                        resourceReferences[rr.RefId],
-                        rr,
-                        incompleteLearningLogItem,
-                        learningLogItemsForResource.Any(ll => ll.CompletedDate != null),
-                        CalculateRecommendedLearningScore(rr, competencyLearningResources, selfAssessmentId, delegateId)
-                    );
-                }
+                rr => GetPopulatedRecommendedResource(
+                    selfAssessmentId,
+                    delegateId,
+                    resourceReferences[rr.RefId],
+                    delegateLearningLogItems,
+                    rr,
+                    competencyLearningResources
+                )
             );
 
-            return recommendedResources;
+            return recommendedResources.Where(r => r != null).Select(r => r!);

```

Comment:
Just a thought, this would be a great extension method if you pulled it out, `WhereNotNull()` or something.